### PR TITLE
fixtures and parametrize

### PR DESCRIPTION
**Pull request by zsoltcsengeri**
Szia István,

A házi feladattal csak félig vagyok kész, próbálom utolérni magam.

Köszönöm,
Zsolt

### CODE REVIEW

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import pytest
+from bankaccount import BankAccount
+
+# Fixture for an empty account
+@pytest.fixture
+def empty_account():
```
Comment (NagyIstvanGabor) [inline homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:6]: itt az empty_account mögött a forrás gondolom az órai anyag, itt azonban üres BankAccount object-et nem tudunk létrehozni. Amint használnád ezt az empty_account fixture-t valahol, hibát kapnál hogy a BankAccount init-je hiányol 2 paramétert, az owner-t és a balance-ot.
Ebben a feladatban egyszerűen létrehozhatsz két különböző BankAccount object-et, pl: bankaccount_1 és bankaccount_2. Az egyiknek az ownere Zsolt, és 100 dollár a balance, a másiknak az ownere "Tamás" és 50 a balance (ez persze csak egy példa, rád vannak bízva a részletek).
Így fogod tudni később tesztelni a transfer metódust a két account között.

**Review by NagyIstvanGabor — CHANGES_REQUESTED** [review]
Szia!

Eddig szuper a házi, egy apró észrevételt jeleztem kommentekben, minden más szuper, várom a folytatást. Köszi szépen!

**Comment by zsoltcsengeri** [comment]
Szia István,

Köszönöm a review-t! Igazából én sem láttam értelmét az üres account-nak, de az órai anyag alapján dolgoztam, és gondoltam, hátha jó lesz valamire. :)
Mindazonáltal most már csináltam két account-ot, ahogy javasoltad, és további teszteseteket is hozzáadtam.

Köszönöm,
Zsolt

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+# Test for bankaccount_1 deposit
+def test_deposit_bankaccount_1(bankaccount_1):
+    bankaccount_1.deposit(600)
+    assert bankaccount_1.balance == 700
+
+
+# Test for bankaccount_2 deposit
+def test_deposit_bankaccount_2(bankaccount_2):
```
Comment (NagyIstvanGabor) [inline homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:33]: Nem hiba, de nem szükséges mindegyik fixture-re írni egy tesztet, ha a deposit logika működik a bankaccount1-re, működni fog a bankaccount2-re is.

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+# Parametrized test for invalid deposit inputs for bankaccount_1
+@pytest.mark.parametrize(
+    "amount, expected_exception",
+    [
+        (-100, ValueError),  # Negative deposit
+        (0, ValueError),  # Zero deposit
+    ],
+)
+def test_deposit_invalid_input(bankaccount_1, amount, expected_exception):
```
Comment (NagyIstvanGabor) [inline homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:46]: ez viszont már annyiból nagyobb gond, hogy ugyanaz a teszt neve ennek is, meg az alatta lévő tesztnek is ahol a bankaccount2-t teszteled. Elég csak az egyik objektumra megírni ezt a tesztet is.

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        (-100, ValueError),  # Negative deposit
+        (0, ValueError),  # Zero deposit
+    ],
+)
+def test_deposit_invalid_input(bankaccount_2, amount, expected_exception):
+    # Test invalid deposit for bankaccount_2
+    with pytest.raises(expected_exception):
+        bankaccount_2.deposit(amount)
+
+    """
```
Comment (NagyIstvanGabor) [inline homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:65]: Ez csak egy apró észrevétel, de ezt az `Edge Cases Test: ` kommentet a test_deposit_invalid_input-ba írtad. Nem okoz semmilyen hibát, de olvashatóság szempontjából tedd egy tabbal kintebbre, akkor egyértelművé válik hogy nem ebből a tesztből maradt le valami, hanem a komment alatti tesztek lesznek az edge case tesztek.

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    Edge Case Test:
+ 
+    """
+
+
+# Transferring money to non-BankAccount objects
+@pytest.mark.parametrize(
```
Comment (NagyIstvanGabor) [inline homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:72]: Itt szuperül megadod a parametrize-ban a paramétereket amikre le kéne fusson a teszt külön külön, viszont végül nem használod a tesztben egyiket sem.
image
Ezen a screenshoton bekarikáztam pirossal az invalid_account paramétert. Látszik, hogy az IDE is jelzi a paraméter színével hogy nem használod sehol ezt az értéket. Ennek a paraméternek a helyére kerülne be az aktuális érték a parametrize listából (először a None, majd a 123, a [] és a {}). 
a 83. sorban tehát a bankaccount_1.transfer()-nek ezt az invalid_account paramétert kéne átadni, így:
`bankaccount_1.transfer(invalid_account)`
Most is lefut 4-szer a teszted ahogy várod, mert a parametrize-ban 4 teszt esetet adsz meg, azonban minden esetben erre a list of dictionary-re fog lefutni a teszt: `[{1, 2, 3, 4, 5}]` mert ezt hard-code-oltad bele a tesztbe.
Mivel TypeError-t vársz mindenhol, még egyszerűbb is lehet a teszted egyébként:

```
@pytest.mark.parametrize(
    "invalid_account",
    [None, 123, [], {}], 
)
def test_transfer_to_invalid_target(invalid_account, bankaccount_1):
    with pytest.raises(TypeError):
        bankaccount_1.transfer(invalid_account, 50)
```

Itt az történik hogy a parametrize listában csak az input értékeket adom meg, ezek bekerülnek minden hívásnál az invalid_account helyére.Mivel a pytest.raises()-ben megadod hogy TypeError-t vársz, így minden teszt esetet erre fog tesztelni.
Át is neveztem a tesztedet transferre, mert alapból deposit-nak hívtad, de a transfer metódust teszteljük. Remélem ez segít kicsit tisztázni a dolgokat, ha van bármi kérdés jelezd kérlek.

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        ({}, TypeError),  # Set
+    ],
+)
+def test_deposit_to_invalid_target(invalid_account, bankaccount_1):
+    with pytest.raises(TypeError):
+        bankaccount_1.transfer([{1, 2, 3, 4, 5}])
+
+
+# Transferring money to the same account
+def test_transfer_to_self(bankaccount_1):
```
Comment (NagyIstvanGabor) [inline homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:87]: ez tökéletes!

**Review by NagyIstvanGabor — CHANGES_REQUESTED** [review]
Szia!

Köszi szépen a második részt, néhány javítani valót jeleztem kommentben részletekkel. Köszönöm!

File: homeworks/zsoltcsengeri/6_testing/test_bankaccount.py (hunk 5 above)
Comment (zsoltcsengeri) [reply homeworks/zsoltcsengeri/6_testing/test_bankaccount.py:72]: Szia István,

Köszönöm az iránymutatást, ismét sokat tanultam belőle.
Elvégeztem a kért változtatásokat, és commitoltam az iránymutatásod alapján.

Üdv,
Zsolt

**Review by NagyIstvanGabor — APPROVED** [review]
Szia!

Köszi szépen a változtatásokat, approve-oltam is. Mehet a merge.
